sound/soc/sof/pcm.c

```diff
@@ -336,21 +340,30 @@ static int sof_pcm_trigger(struct snd_soc_component *component,
 		return -EINVAL;
 	}
 
-	/*
-	 * DMA and IPC sequence is different for start and stop. Need to send
-	 * STOP IPC before stop DMA
-	 */
-	if (!ipc_first)
+	if (!pcm_ops_first)
 		snd_sof_pcm_platform_trigger(sdev, substream, cmd);
 
 	if (pcm_ops && pcm_ops->trigger)
 		ret = pcm_ops->trigger(component, substream, cmd);
 
-	/* need to STOP DMA even if trigger IPC failed */
-	if (ipc_first)
+	switch (cmd) {
+	case SNDRV_PCM_TRIGGER_PAUSE_RELEASE:
+	case SNDRV_PCM_TRIGGER_START:
+		/* invoke platform trigger to start DMA only if pcm_ops is successful */
+		if (pcm_ops_first && !ret)
+			snd_sof_pcm_platform_trigger(sdev, substream, cmd);
+		break;
+	case SNDRV_PCM_TRIGGER_SUSPEND:
+	case SNDRV_PCM_TRIGGER_PAUSE_PUSH:
+	case SNDRV_PCM_TRIGGER_STOP:
+		/* invoke platform trigger to stop DMA even if pcm_ops failed */
 		snd_sof_pcm_platform_trigger(sdev, substream, cmd);
+		break;
+	default:
+		break;
+	}
```

Comment:
wondering if there should be a first patch that renames ipc_first as pcm_ops_first and adds the test to  "start DMA only if pcm_ops is successful", and a second one where you add the trigger_ipc_first_on_start change specific to IPC4

It feels like you are changing too many things at once.

Reply:
yes, I've dropped the change in the bool variable name.

Reply:
there's still a big change with this switch() block and a change of behavior with  "start DMA only if pcm_ops is successful "

I am just reasonably worried that we should not compound two changes in one patch for non-regressions tests.